[1.20.x] Optimise path filter in mod loading

PaintNinja opened the pull request:
Port from downstream.

Setting these to null instead of a lambda that always returns true enables two optimisations which skips a bunch of unnecessary code:
https://github.com/McModLauncher/securejarhandler/blob/1dedb0655d9f4b8cf310ea7095b38c7a978a42ab/src/main/java/cpw/mods/niofs/union/UnionFileSystem.java#L397
and
https://github.com/McModLauncher/securejarhandler/blob/1dedb0655d9f4b8cf310ea7095b38c7a978a42ab/src/main/java/cpw/mods/jarhandling/SecureJar.java#L90

PaintNinja committed:
Optimise path filter during mod loading

Partial port from downstream, excluding a possible breaking change.

Co-Authored-By: embeddedt <42941056+embeddedt@users.noreply.github.com>

Jonathing commented:
This PR changes the functions to null but, like you said in your comment, doesn't account for null checking right now on filtering. Is it worth merging this one?

Jonathing reviewed (approved):
Nevermind, this is fine. I don't know enough about how ModLauncher or SJH work, but I trust you've done the necessary testing to make it work.

PaintNinja commented:
The ones I've done go through the same kind of path when jumping through using your IDE...
`AbstractModProvider.createMod()` calls `SecureJar.from()` with the `null` filter, which calls the `Jar` constructor, which calls `UFSP.newFileSystem()` with the path filter, which calls `newFileSystemInternal()`, which creates a new `UnionFileSystem` with the path filter, which sets it. If you check the public `UnionFileSystem.getFilesystemFilter()`, it's called by `Jar.getProviders()` where it's fed into `SecureJar.Provider.fromPath()` where it does the null check.

Looked into this more closely and I derped, mixed up `CommonLaunchHandler.minecraftPaths()` and `LocatedPaths.minecraftFilter()`. The filter is what's being changed to null and that goes through `ModJarMetadata.buildFile()` which then goes down the usual `SecureJar.from()` chain explained earlier which contains the null check.

I'll add those back in now

PaintNinja committed:
Add in other changes

I was wrong, these shouldn't be breaking changes.

LexManos commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java:
```diff
@@ -38,7 +38,7 @@ protected IModLocator.ModFileOrException createMod(Path... path) {
         var sj = SecureJar.from(
                 Manifest::new,
                 jar -> jar.moduleDataProvider().findFile(MODS_TOML).isPresent() ? mjm : JarMetadata.from(jar, path),
-                (root, p) -> true,
```
This should just delete the line, to call the variant of from that doesn't take the BiPredicate. It is functionally the same, but from an API usage point it avoids a 'magic null'

LexManos commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModJarMetadata.java:
```diff
@@ -31,7 +31,7 @@ static Optional<IModFile> buildFile(IModLocator locator, Predicate<SecureJar> ja
     // TODO: Remove helper functions to cleanup api
     @Deprecated(forRemoval = true, since="1.18")
```
So apparently our deprecation validator doesn't test fmlloader. I will need to add that to the list of things to fix.

LexManos commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/targets/CommonClientLaunchHandler.java:
```diff
@@ -40,7 +40,7 @@ public LocatedPaths getMinecraftPaths() {
         var lowcodelang = LibraryFinder.findPathForMaven(vers.forgeGroup(), "lowcodelanguage", "", "", vers.mcAndForgeVersion());
         var mclang = LibraryFinder.findPathForMaven(vers.forgeGroup(), "mclanguage", "", "", vers.mcAndForgeVersion());
 
-        return new LocatedPaths(mcstream.build().toList(), (a,b) -> true, modstream.build().toList(), List.of(fmlcore, javafmllang, lowcodelang, mclang));
```
LocatedPaths could use a @Nullable if you're adding this, but besides that looks fine.

PaintNinja replied:
I considered that, however I decided against it as it would either imply everything else isn't nullable or mean that I'd have to comb through a much larger chunk of code to annotate stuff properly.

LexManos commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/targets/CommonServerLaunchHandler.java:
```diff
@@ -39,13 +39,21 @@ public LocatedPaths getMinecraftPaths() {
             }, mcextra
         );
+
+        // Todo in 1.20.2: Change the default filter to null instead of always true.
```
We really should consider all the LaunchHandlers as internal API. But yes I can see this backwards compatibility for backports.

PaintNinja replied:
Now that 1.20.2 is out, should I remove the back-compat in this PR but keep it in the backports? Or keep it as-is here too?

LexManos replied:
Go ahead and break it, considering what I said about this basically being internal API. And our only implementation of it being a case that doesn't care.

PaintNinja replied:
Done

LexManos reviewed (approved):
A few things that could be cleaned up but not a big deal. More importantly added a few things to my todo.

PaintNinja committed:
Merge branch 'MinecraftForge:1.20.x' into 1-20-optimise-path-filter-during-mod-loading

PaintNinja committed:
Use method variant that doesn't take the BiPredicate

Functionally the same, but slightly cleaner API usage

PaintNinja committed:
Minor breaking change to `CommonServerLaunchHandler`

net.minecraftforge.fml.loading.targets.CommonServerLaunchHandler#processMCStream's filter is now null by default instead of an always-true bipredicate.
